cgi/import_file_upload.pl

```diff
@@ -0,0 +1,240 @@
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use Modern::Perl '2017';
+use utf8;
+
+binmode(STDOUT, ":encoding(UTF-8)");
+binmode(STDERR, ":encoding(UTF-8)");
+
+use CGI::Carp qw(fatalsToBrowser);
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Store qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Users qw/:all/;
+use ProductOpener::Images qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Mail qw/:all/;
+
+use Apache2::RequestRec ();
+use Apache2::Const ();
+
+use CGI qw/:cgi :form escapeHTML :cgi-lib/;
+use URI::Escape::XS;
+use Storable qw/dclone/;
+use Encode;
+use JSON::PP;
+use Log::Any qw($log);
+
+ProductOpener::Display::init();
+
+my $type = param('type') || 'upload';
+my $action = param('action') || 'display';
+
+my $title = '';
+my $html = '';
+
+local $log->context->{type} = $type;
+local $log->context->{action} = $action;
+
+if (not defined $owner) {
+	display_error(lang("no_owner_defined"), 200);
+}
+
+if ($action eq "process") {
+
+	# Process uploaded files
+
+	my $file = param('file_input_data');
+
+	my %data = ();
+
+	if ($file =~ /\.(xlsx|csv|tsv)$/i) {
+
+
+		my $extension = lc($1) ;
+		my $filename = $`;
+		my $uploaded_t = time();
+		my $file_id = $uploaded_t . '-' . get_string_id_for_lang("no_language", $filename);
+
+		$log->debug("processing upload form", { filename => $filename, file_id => $file_id, extension => $extension }) if $log->is_debug();
+
+		(-e "$data_root/import_files") or mkdir("$data_root/import_files", 0755);
+		(-e "$data_root/import_files/$owner") or mkdir("$data_root/import_files/$owner", 0755);
+
+		open (my $out, ">", "$data_root/import_files/$owner/$file_id.$extension") ;
+		while (my $chunk = <$file>) {
+			print $out $chunk;
+		}
+		close ($out);
+
+		%data = (
+			location => "$formatted_subdomain/cgi/import_file_select_format.pl?file_id=$file_id&action=display",
+		);
+
+		# Keep track of uploaded files attributes and status
+
+		my $import_files_ref = retrieve("$data_root/import_files/$owner/import_files.sto");
+		if (not defined $import_files_ref) {
+			$import_files_ref = {};
+		}
+
+		$import_files_ref->{$file_id} = {
+			filename => $filename,
+			extension => $extension,
+			uploaded_t => $uploaded_t,
+		};
+
+		store("$data_root/import_files/$owner/import_files.sto", $import_files_ref);
```

Comment:
Do we want to delete old uploads at some point in time?

Reply:
Probably at some point, but at the start at least, I prefer to keep everything, including temporary files etc. so that it's easier to debug issues. We can always have a batch process to clean things periodically.